daml/ContingentClaims/Lifecycle.daml

```diff
@@ -75,39 +81,51 @@ lifecycle spot
     foldE (Right x) = x
 
 -- | Evaluate observables and skip branches for which predicates don't hold.
--- Consumes `When` and `Cond` when appropriate, leaving their subtrees.
 -- This is useless on its own. Composed with other functions, it adds laziness.
 acquire' : (Ord t, Eq a, Monad m)
   => (a -> t -> m Decimal)
   -> C t a -> Kleisli m t (F t a (Either (C t a) (C t a)))
 acquire' spot (When obs c) = do
   predicate <- compare spot obs
-  if predicate then acquire' spot c else pure $ WhenF obs (Left c)
+  pure $ WhenF obs if predicate then Right c else Left c
 acquire' spot (Cond obs c c') = do
   predicate <- compare spot obs
-  if predicate then acquire' spot c else acquire' spot c'
+  pure if predicate 
+    then CondF obs (Right c) (Left c') 
+    else CondF obs (Left c) (Right c')
 acquire' spot (Anytime obs c) = do
   predicate <- compare spot obs
   pure $ AnytimeF obs if predicate then Right c else Left c
 acquire' _ other = pure $ Right <$> project other
 
--- | Evaluate any observables in `Scale` nodes, accumulating scale factors
--- top-down, and log these with their corresponding leaf values. Skip `Or` and
--- `Anytime` branches, guaranteeing liveness.
-lifecycle' : CanAbort m
+-- | Evaluate any observables in `Scale` nodes, accumulating scale factors and 
+-- propagating the claim's acquisition times top-down.
+-- Log the scale factors with their corresponding leaf values. 
+-- Skip `Or` and `Anytime` branches, guaranteeing liveness. 
+-- Replace any Ones that can be reached with Zeros.
+lifecycle' : (Ord t, CanAbort m)
         => (a -> t -> m Decimal)
-        -> (Decimal, C t a) -> WriterT [(Decimal, a)] (Kleisli m t) (F t a (Either (C t a) (Decimal, C t a)))
-lifecycle' _ (qty, One asset) = do
+        -> ((t, Decimal), C t a) -> WriterT [(Decimal, a)] (Kleisli m t) (F t a (Either (C t a) ((t, Decimal), C t a)))
+lifecycle' _ ((_, qty), One asset) = do
   tell [(qty, asset)]
   pure $ Right <$> ZeroF
-lifecycle' _ (qty, Give c) =
-  pure . GiveF $ Right (-qty, c)
-lifecycle' spot (qty, Scale obs c) = do
-  k <- lift $ Observation.eval spot obs
-  pure . ScaleF obs $ Right (k * qty, c)
+lifecycle' _ ((s, qty), Give c) =
+  pure . GiveF $ Right ((s, -qty), c)
+lifecycle' spot ((s, qty), Scale obs c) = do
+  k <- lift $ Kleisli \ _ -> runKleisli (Observation.eval spot obs) s
```

Comment:
The observable is evaluated at the acquisition time `s` instead of the today date `t`

Reply:
I also need to review this, but I think that, as we need to evaluate the predicates (observables) as we traverse the tree, it doesn't make sense to return a `Kleisli` function at all. 
Previously I was missing this point.
This is made evident by the fact that you're throwing away the `t`. i.e. `Kleisli \_t -> ...`
I'd encourage you to try and remove the `Kleisli` from the type signature altogether (and that may propagate down to observations). I think it will considerably simplify the code.

The `lifecycle` signature won't change, but the implementation will.

Reply:
Thanks for the feedback, I just pushed a commit that removes the Kleisli wrapper.